### ISSUE TITLE
Switch the API away from Borrow bounds

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#2500]
It is a very common pattern in gfx-hal API right now to have `I: IntoIterator` and `I::Item: Borrow<Something>`. Unfortunately, this falls apart if the user actually wants to have something like this:
```rust
struct Foo<B: Backend>(B::Something);
impl<B: Backend> Borrow<B::Something> for Foo<B> {...}
```
Compiler isn't happy, because `T` impls `Borrow<T>` and there is no way to convince the compiler that this implementation doesn't conflict with our new one...

I'm going to change everything to use `AsRef` therefore. Any thoughts? Objections?

**Comment by omni-viral** [comment]
There is a way to convince the compiler

```
struct Foo<B, S=B::Something>(S);
impl<B> Borrow<B::Something> for Foo<B> { ... }
```

**Comment by kvark** [comment]
Apparently, `AsRef` isn't a solution either, since it's not implemented for `T` or `&T`. Next alternative: `Iterator<&'a T>`

**Comment by omni-viral** [comment]
`Iterator<Item = &'a T>` isn't better than `Iterator<Item = impl Borrow<T>>`

**Comment by kvark** [comment]
@omni-viral it *is* better because `Borrow<T>` just isn't possible for structs that just depend on `<B: Backend>`, including our own `CommandBuffer`. The trade-off here is that for `Borrow` we'd require all the users to change their data types/structures, while for `&'a T` it's only a single `iter()` call (or similar) at the use site.

**Comment by kvark** [comment]
Interestingly, `Iterator<Item = &'a T>` isn't a solution either :( If the user has an iterator over actual items, I don't see how Rust would allow us to make an iterator over their references instead (sounds like a case for streaming iterators, which we don't have). See [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d0f9d491bb9483041d1d6fd4710ad6ac).

TL;DR: no solution is good atm?
  - `Borrow` is bad because we can't `impl<B: Backend> Borrow<B::AssociatedType> for Something<B>`, where `Something<B: Backend>(B::AssociatedType)` or similar
    - a workaround exist by adding an extra generic type, but it's a bit awkward
  - `AsRef` is bad because it's not implemented for `T` or `&T`
  - `&'a T` is bad because we can't get an iterator over references if we have one over owned items...

@Gankro - maybe you can make a recommendation for us?

**Comment by Gankro** [comment]
This stuff is extremely out of cache for me, but yeah in general you can't "win" the generic borrow game. The ideal system has conflicting blanket impls, and is consequently inexpressible. I also don't know enough about your system to suggest a more subtle solution. The only advice I can give is "trying to give a million meta-conversions to smooth over APIs is something I dislike as it blows up compile times and makes it much harder to read signatures, so maybe don't do it?"

**Comment by omni-viral** [comment]
@kvark you missed my point. `impl Borrow<T>` matches both `&T` and `T` already. So it is already better than just plain `&T`. Even though you not always can implement it for wrapper type (although you  can)

**Comment by termhn** [comment]
> "trying to give a million meta-conversions to smooth over APIs is something I dislike as it blows up compile times and makes it much harder to read signatures, so maybe don't do it?"

Agree

**Comment by omni-viral** [comment]
The other option is to not use `B: Backend` as type parameter for type wrappers.
So instead of 
```rust
struct Foo<B: Backend> {
  raw: B::Foo,
}
```

you'd write

``rust
struct Foo<F> {
  raw: F,
}

impl<F> Borrow<F> for Foo<F> { ... }
```

Then function signature
```rust

trait Device<B: Backend> {
  fn takes_foos(_: impl IntoIterator<Item = impl Borrow<B::Foo>>);
}

// This should typecheck.
let foos: Vec<Foo<B::Foo>> = ...;
device.takes_foos(foos);
device.takes_foos(&foos);
```

**Comment by Gankro** [comment]
I wrote up a detailed explanation here of why you can logically reason about the non-existence of the overlapping implementation, but in fact it can exist, and Rust doesn't factor in this kind of "local non-existence": https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b0e9517c0599b080f9eeae3bfe548aa2

**Comment by Gankro** [comment]
Oh also, although streaming iterators aren't compatible with iterators/for, you can make them, they work fine: https://docs.rs/streaming-iterator/0.1.4/streaming_iterator/

**Comment by kvark** [comment]
@Gankro @omni-viral I'm really puzzled as to why this gets a pass:
```rust
trait Bar {
    type Something;
}
struct Foo<B: Bar, S=<B as Bar>::Something>(S, PhantomData<B>);
impl<B, S> Borrow<S> for Foo<B, S>
where
    B: Bar<Something = S>,
{
    fn borrow(&self) -> &B::Something {
        &self.0
    }
}
```

Isn't saying that `B<Something = S>` semantically equivalent to just using `B::Something` instead of `S` in the first place? :/

**Comment by Gankro** [comment]
filed https://github.com/rust-lang/rust/issues/56804 just to be sure this isn't "accidentally" working

**Comment by omni-viral** [comment]
@kvark 
The sole fact that `S` is type parameter to `Foo<B, S>` convince rustc that `Foo<B, S> != S` hence 
`Foo<B> != B::Something`.

**Comment by omni-viral** [comment]
Which make it possible to implement `Borrow<B::Something>` without collision.

**Comment by madadam** [comment]
Could this be solved by a custom `Borrow`-like trait:

```rust
pub trait GfxBorrow<T> {
    fn borrow(&self) -> &T;
}
```
And manually implemented it for all the types for which it makes sense?

**Comment by kvark** [comment]
@madadam the point of `Borrow` bound here (or any alternative to it) is to support the user having their own structures as wrappers around gfx-rs's raw types. So if we introduce our own trait, we'd be forcing the users to implement it, which would be unfortunate.

**Comment by madadam** [comment]
Sure, but is that different from the situation now, where they have to implement `Borrow`?

**Comment by kvark** [comment]
@madadam yeah, I suppose that is mostly the case. `Borrow` is usefully implemented for a bunch of standard types, such as smart pointers, so I thought it could be handy.

**Comment by madadam** [comment]
The custom trait can be implemented for those types too. The idea is just to explicitly not implement it for `T`, to avoid the original issue here.

**Comment by kvark** [comment]
Funny thing is - we *do* need that blanket implementation for `T` (i.e. provide actual `&B::CommandBuffer` not a wrapper around it). So we end up with the same problem even having a custom trait :)

**Comment by madadam** [comment]
No :) What I'm trying to suggest is:

1. Implement the custom trait **manually** for each gfx type
2. Implement it additionally for some useful types like `Arc`, `Rc`, ...
3. Let the users implement it themselves for their own wrappers

Point 1 is tedious, but it could be handled by a macro or custom derive. We could even provide custom derive to handle point 3, but that is probably overkill.

**Comment by kvark** [comment]
We have a `CommandBuffer` type wrapper in gfx-hal (hello #2514) that wraps around `B::CommandBuffer`. Even if we introduce a new trait `MyBorrow<T>` (which I tried), it would have to be implemented for both `CommandBuffer<B>` and `B::CommandBuffer`. And it's impossible, it faces the same issue as with the original `Borrow`.

**Comment by madadam** [comment]
I see. Then I think I'm out of ideas :slightly_frowning_face:

**Comment by kvark** [comment]
This would be deprecated by #2862
